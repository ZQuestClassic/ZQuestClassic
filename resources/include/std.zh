// std.zh
// 28th December, 2020 for ZC 2.55 Alpha 86
// Standard header for ZScript.
// Please don't edit this file. Instead, create a new header file for your
// quests' scripts,  add constants and procedures to that file,
// and #include it separately.

//Options
#option BINARY_32BIT off
	//Would break a number of things if set on
#option SHORT_CIRCUIT on
	//Should work for everything in std_zh, and just speed it up

#includepath "include"

//Version Metadata
#include "std_zh/std_meta.zh"

//User Configuration Files
#include "std_zh/std_user_defs.zh" //Typedefs: Always first.
#include "std_zh/std.cfg"

//Main Files
#include "std_zh/limits.zh"
#include "std_zh/std_constants.zh"
#include "std_zh/ffrules.zh"
#include "std_zh/std_functions.zh"
#include "std_zh/math.zh"
#include "std_zh/weapon.zh"
#include "std_zh/std_sideview.zh" //Sideview and platforming

//ZScript string handling functions, akin to those found in C
//#include "string.zh"
#include "std_zh/string/string_constants.zh" //Now #includeed from here. This prevents conflicts for those people
#include "std_zh/string/string_functions.zh" //who would #include string.zh from a secondary source, as that file is now 
			     //empty, and the string.zh components load from here. 

//Keyboard handling for 2.54 and above
#include "std_zh/std_keyboard.zh"

//Locale
#include "std_zh/locale/std_EnglishIntl.zh" //International English Localisation

//Legacy Wrappers
#include "std_zh/std_legacy.zh"