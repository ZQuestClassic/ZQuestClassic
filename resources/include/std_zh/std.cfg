//Settings

const int STD_FORCE_OLD_DISTANCE = 1;	//If enabled, Distance() will use Sqrt() instead of SafeSqrt().
					//! If you use this, disable the quest rule 'Log Script Errors to Allegro.log
					//! before distributing your quest!
					
const int STD_DISTANCE_USE_SAFESQRT = 0; 	//If enebaled, Distance will use SafeSqrt(). Otherwise, it will return a factor, 
					
const int STD_EMULATE_ENGINE_COLLISION = 0; // If enabled, Collision hitboxs will more precisely emulate the
					    // internal collision. This differs to how scripted collision
					    // has always worked.
					    

const int STD_CENTERXY_USES_HITBOXES = 0; // If enabled, CenterX() and CenterY() will use object hitboxes
					  // and offsets by default, rather than sprite sizes when calculating
					  // the centre of objects.
					  

const int STD_NEVER_USE_SAFESQRT = 1; //Enable to force any calls to SafeSqrt() to use Sqrt() instead.

const int MAX_HIGHEST_LEVEL_ITEM_CHECK = 255; //The highest item ID checked by GetHighestLevelItem(). 

const int REVISED_ROTATEPOINT_X_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointX()
const int REVISED_ROTATEPOINT_Y_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointY()


const int MAX_NOACTION_INPUT = 14; //Checks A, B, L, R, U/D/L/R, Ex1, Ex2, Ex3, Ex4
	//Use CB_MAX to force the analogue stick false. 
const int NOACTION_SKIP_MAP = 0; //If enabled, NoAction will not disable pressing the Map button
const int NOACTION_SKIP_START = 0; //If enabled, NoAction will not disable pressing the Start button

const int OLD_ONSIDEPLATFORM = 0; //Change to 1 to use old OnSidePlatform calls.

const int OLD_RAND_FUNCS = 0; //Change to 1 to use old Rand() logic
