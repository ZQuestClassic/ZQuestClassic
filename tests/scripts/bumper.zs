#include "std.zh"

const int LTTP_BUMPER_FORCE = 3; //How fast Link gets pushed back when he hits a bumper
const int LTTP_BUMPER_ANIM_SPEED = 4; //How fast the bumpers animate in frames
const int SFX_LTTP_BUMPER = 76; //The sound that plays when Link hits a bumper

const int MAX_PUSH = 4; //Max speed the script can move Link in one frame

//Combo Setup: Each set of bumper combos should be arranged as so: 
//Bumper Frame 1
//Bumper Frame 2
//Bumper Frame 3
//Bumper Frame 4
//Each of the four combos is a frame in the bumper's animation.
//When Link hits the bumper it changes to frame 4 and then goes back a combo every LTTP_BUMPER_ANIM_SPEED frames.
//The FFC should use the first of the four bumper combos.
//D0: How many frames Link gets pushed back for when he hits the bumper

ffc script LttP_Bumper{
	void run(int Bounce){
		int Combo = this->Data;
		int BounceAngle = 0;
		int BounceCounter = 0;
		this->InitD[7] = 0;
		int AnimationCounter = 0;
		int PushX;
		int PushY;
		while(true){
			if(Distance(CenterX(this), CenterY(this), CenterLinkX(), CenterLinkY())<this->TileWidth*8+2&&Link->Z==0){
				if(this->InitD[7]==0){
					Audio->PlaySound(SFX_LTTP_BUMPER);
					BounceAngle = Angle(CenterX(this), CenterY(this), CenterLinkX(), CenterLinkY());
					BounceCounter = Bounce;
					AnimationCounter = LTTP_BUMPER_ANIM_SPEED*4;
				}
				NoAction();
			}
			if(BounceCounter>0){
				PushX += VectorX(LTTP_BUMPER_FORCE, BounceAngle);
				PushY += VectorY(LTTP_BUMPER_FORCE, BounceAngle);
				BounceCounter--;
			}
			if(Abs(PushX)>0){
				for(int i=0; i<MAX_PUSH&&PushX<=-1; i++){
					if(CanWalk(Link->X, Link->Y, DIR_LEFT, 1, false)){
						Link->X--;
						PushX++;
					}
					else{
						if(Abs(PushY)>0)
							PushY += Sign(PushY)*Abs(PushX/2);
						PushX = 0;
					}
				}
				for(int i=0; i<MAX_PUSH&&PushX>=1; i++){
					if(CanWalk(Link->X, Link->Y, DIR_RIGHT, 1, false)){
						Link->X++;
						PushX--;
					}
					else{
						if(Abs(PushY)>0)
							PushY += Sign(PushY)*Abs(PushX/2);
						PushX = 0;
					}
				}
			}
			if(Abs(PushY)>0){
				for(int i=0; i<MAX_PUSH&&PushY<=-1; i++){
					if(CanWalk(Link->X, Link->Y, DIR_UP, 1, false)){
						Link->Y--;
						PushY++;
					}
					else{
						if(Abs(PushX)>0)
							PushX += Sign(PushX)*Abs(PushY/2);
						PushY = 0;
					}
				}
				for(int i=0; i<MAX_PUSH&&PushY>=1; i++){
					if(CanWalk(Link->X, Link->Y, DIR_DOWN, 1, false)){
						Link->Y++;
						PushY--;
					}
					else{
						if(Abs(PushX)>0)
							PushX += Sign(PushX)*Abs(PushY/2);
						PushY = 0;
					}
				}
			}
			if(AnimationCounter>0)
				AnimationCounter--;
			this->Data = Combo+Floor(AnimationCounter/LTTP_BUMPER_ANIM_SPEED);
			Waitframe();
		}
	}
}