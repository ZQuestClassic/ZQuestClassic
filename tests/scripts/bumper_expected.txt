Compiling 'bumper.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Generating object code
Pass 6: Assembling
Success!
Compile finished with exit code '0' (success)

PUSHR d0
PUSHR REFFFC
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
SETR d4,SP
LOADD d2,7
PUSHR d2
POP REFFFC
SETR d2,DATA
STORED d2,6
SETV d2,0
STORED d2,5
SETV d2,0
STORED d2,4
SETV d2,0
PUSHR d2
LOADD d2,7
POP d3
PUSHR d2
PUSHR d3
SETV d2,7
PUSHR d2
POP d0
POP d2
POP REFFFC
SETR FFINITDD,d2
SETV d2,0
STORED d2,3
SETV d2,0
STORED d2,2
SETV d2,0
STORED d2,1
PUSHR d4
PUSHV 0.0065
PUSHR d4
PUSHV 0.0045
LOADD d2,7
PUSHR d2
GOTO 917
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0052
LOADD d2,7
PUSHR d2
GOTO 941
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0057
GOTO 907
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0062
GOTO 912
POP d4
PUSHR d2
GOTO 517
POP d4
PUSHR d2
LOADD d2,7
PUSHR d2
POP REFFFC
SETR d2,FFTWIDTH
MULTV d2,8
ADDV d2,2
POP d3
COMPARER d3,d2
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 84
SETR d2,LINKZ
COMPAREV d2,0
SETTRUEI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 138
LOADD d2,7
PUSHR d2
SETV d2,7
PUSHR d2
POP d0
POP REFFFC
SETR d2,FFINITDD
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 134
SETV d2,76
PLAYSOUNDR d2
PUSHR d4
PUSHV 0.0128
PUSHR d4
PUSHV 0.0108
LOADD d2,7
PUSHR d2
GOTO 917
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0115
LOADD d2,7
PUSHR d2
GOTO 941
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0120
GOTO 907
POP d4
PUSHR d2
PUSHR d4
PUSHV 0.0125
GOTO 912
POP d4
PUSHR d2
GOTO 580
POP d4
STORED d2,5
LOADD d2,8
STORED d2,4
SETV d2,16
STORED d2,3
PUSHR d4
PUSHV 0.0137
GOTO 965
POP d4
LOADD d2,4
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 174
LOADD d2,2
PUSHR d2
PUSHR d4
PUSHV 0.0153
PUSHV 3
LOADD d2,5
PUSHR d2
GOTO 1010
POP d4
POP d3
ADDR d2,d3
STORED d2,2
LOADD d2,1
PUSHR d2
PUSHR d4
PUSHV 0.0165
PUSHV 3
LOADD d2,5
PUSHR d2
GOTO 1021
POP d4
POP d3
ADDR d2,d3
STORED d2,1
LOADD d2,4
PUSHR d2
SUBV d2,1
STORED d2,4
POP d2
LOADD d2,2
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 330
SETV d2,0
STORED d2,0
LOADD d2,0
COMPAREV d2,4
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 195
LOADD d2,2
COMPAREV d2,-1
SETLESSI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 256
PUSHR d4
PUSHV 0.0207
PUSHR LINKX
PUSHR LINKY
PUSHV 2
PUSHV 1
PUSHV 0
GOTO 600
POP d4
COMPAREV d2,0
GOTOTRUE 223
SETR d2,LINKX
PUSHR d2
SUBV d2,1
PUSHR d2
POP d2
SETR LINKX,d2
POP d2
LOADD d2,2
PUSHR d2
ADDV d2,1
STORED d2,2
POP d2
GOTO 250
LOADD d2,1
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 248
LOADD d2,1
PUSHR d2
PUSHR d4
PUSHV 0.0238
LOADD d2,1
PUSHR d2
GOTO 1032
POP d4
PUSHR d2
LOADD d2,2
DIVV d2,2
ABS d2
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
STORED d2,1
SETV d2,0
STORED d2,2
LOADD d2,0
PUSHR d2
ADDV d2,1
STORED d2,0
POP d2
GOTO 184
SETV d2,0
STORED d2,0
LOADD d2,0
COMPAREV d2,4
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 269
LOADD d2,2
COMPAREV d2,1
SETMOREI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 330
PUSHR d4
PUSHV 0.0281
PUSHR LINKX
PUSHR LINKY
PUSHV 3
PUSHV 1
PUSHV 0
GOTO 600
POP d4
COMPAREV d2,0
GOTOTRUE 297
SETR d2,LINKX
PUSHR d2
ADDV d2,1
PUSHR d2
POP d2
SETR LINKX,d2
POP d2
LOADD d2,2
PUSHR d2
SUBV d2,1
STORED d2,2
POP d2
GOTO 324
LOADD d2,1
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 322
LOADD d2,1
PUSHR d2
PUSHR d4
PUSHV 0.0312
LOADD d2,1
PUSHR d2
GOTO 1032
POP d4
PUSHR d2
LOADD d2,2
DIVV d2,2
ABS d2
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
STORED d2,1
SETV d2,0
STORED d2,2
LOADD d2,0
PUSHR d2
ADDV d2,1
STORED d2,0
POP d2
GOTO 258
LOADD d2,1
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 486
SETV d2,0
STORED d2,0
LOADD d2,0
COMPAREV d2,4
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 351
LOADD d2,1
COMPAREV d2,-1
SETLESSI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 412
PUSHR d4
PUSHV 0.0363
PUSHR LINKX
PUSHR LINKY
PUSHV 0
PUSHV 1
PUSHV 0
GOTO 600
POP d4
COMPAREV d2,0
GOTOTRUE 379
SETR d2,LINKY
PUSHR d2
SUBV d2,1
PUSHR d2
POP d2
SETR LINKY,d2
POP d2
LOADD d2,1
PUSHR d2
ADDV d2,1
STORED d2,1
POP d2
GOTO 406
LOADD d2,2
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 404
LOADD d2,2
PUSHR d2
PUSHR d4
PUSHV 0.0394
LOADD d2,2
PUSHR d2
GOTO 1032
POP d4
PUSHR d2
LOADD d2,1
DIVV d2,2
ABS d2
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
STORED d2,2
SETV d2,0
STORED d2,1
LOADD d2,0
PUSHR d2
ADDV d2,1
STORED d2,0
POP d2
GOTO 340
SETV d2,0
STORED d2,0
LOADD d2,0
COMPAREV d2,4
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 425
LOADD d2,1
COMPAREV d2,1
SETMOREI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 486
PUSHR d4
PUSHV 0.0437
PUSHR LINKX
PUSHR LINKY
PUSHV 1
PUSHV 1
PUSHV 0
GOTO 600
POP d4
COMPAREV d2,0
GOTOTRUE 453
SETR d2,LINKY
PUSHR d2
ADDV d2,1
PUSHR d2
POP d2
SETR LINKY,d2
POP d2
LOADD d2,1
PUSHR d2
SUBV d2,1
STORED d2,1
POP d2
GOTO 480
LOADD d2,2
ABS d2
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 478
LOADD d2,2
PUSHR d2
PUSHR d4
PUSHV 0.0468
LOADD d2,2
PUSHR d2
GOTO 1032
POP d4
PUSHR d2
LOADD d2,1
DIVV d2,2
ABS d2
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
STORED d2,2
SETV d2,0
STORED d2,1
LOADD d2,0
PUSHR d2
ADDV d2,1
STORED d2,0
POP d2
GOTO 414
LOADD d2,3
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 498
LOADD d2,3
PUSHR d2
SUBV d2,1
STORED d2,3
POP d2
LOADD d2,6
PUSHR d2
LOADD d2,3
DIVV d2,4
FLOOR d2
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,7
POP d3
PUSHR d2
PUSHR d3
POP d2
POP REFFFC
SETR DATA,d2
WAITFRAME
GOTO 38
POPARGS d5,0.0009
QUIT
POP d2
POP d1
POP d0
POP d6
SETR d2,DISTANCE
RETURN
SETR d4,SP
LOADD d2,0
COMPAREV d2,8
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0.0001
GOTOMORE 538
LOADD d2,0
COMPAREV d2,15
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
GOTOTRUE 543
LOADD d2,0
GOTO 578
LOADD d2,0
SETR SWITCHKEY,d2
COMPAREV SWITCHKEY,8
GOTOTRUE 562
COMPAREV SWITCHKEY,9
GOTOTRUE 564
COMPAREV SWITCHKEY,10
GOTOTRUE 566
COMPAREV SWITCHKEY,11
GOTOTRUE 568
COMPAREV SWITCHKEY,12
GOTOTRUE 570
COMPAREV SWITCHKEY,13
GOTOTRUE 572
COMPAREV SWITCHKEY,14
GOTOTRUE 574
COMPAREV SWITCHKEY,15
GOTOTRUE 576
GOTO 578
SETV d2,0
GOTO 578
SETV d2,5
GOTO 578
SETV d2,3
GOTO 578
SETV d2,7
GOTO 578
SETV d2,1
GOTO 578
SETV d2,6
GOTO 578
SETV d2,2
GOTO 578
SETV d2,4
GOTO 578
POPARGS d5,0.0001
RETURN
SETR d4,SP
LOADD d2,1
PUSHR d2
LOADD d2,3
POP d3
SUBR d3,d2
PUSHR d3
LOADD d2,0
PUSHR d2
LOADD d2,2
POP d3
SUBR d3,d2
SETR d1,d3
POP d0
ARCTANR d2
SETR d3,d2
RADTODEG d2,d3
GOTO 598
POPARGS d5,0.0004
RETURN
PUSHR d2
PUSHR d2
PUSHR d2
PUSHR d2
SETR d4,SP
SETV d2,8
STORED d2,3
LOADD d2,8
ADDV d2,15
STORED d2,2
LOADD d2,7
ADDV d2,15
STORED d2,1
LOADD d2,4
COMPAREV d2,0
GOTOTRUE 618
SETV d2,0
STORED d2,3
PUSHR d4
PUSHV 0.0623
LOADD d2,6
PUSHR d2
GOTO 523
POP d4
SETR SWITCHKEY,d2
COMPAREV SWITCHKEY,0
GOTOTRUE 634
COMPAREV SWITCHKEY,1
GOTOTRUE 705
COMPAREV SWITCHKEY,2
GOTOTRUE 761
COMPAREV SWITCHKEY,3
GOTOTRUE 828
GOTO 892
LOADD d2,7
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
SETR d2,d3
COMPAREV d2,0
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0.0001
GOTOMORE 662
LOADD d2,8
PUSHR d2
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
SETR d1,d3
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 682
LOADD d2,8
ADDV d2,8
PUSHR d2
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
SETR d1,d3
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 701
LOADD d2,2
PUSHR d2
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
SETR d1,d3
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
SETTRUEI d2
GOTO 905
LOADD d2,1
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
COMPAREV d2,176
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 726
LOADD d2,8
PUSHR d2
LOADD d2,1
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 742
LOADD d2,8
ADDV d2,8
PUSHR d2
LOADD d2,1
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 757
LOADD d2,2
PUSHR d2
LOADD d2,1
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
SETTRUEI d2
GOTO 905
LOADD d2,8
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
SETR d2,d3
COMPAREV d2,0
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0.0001
GOTOMORE 789
LOADD d2,8
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
PUSHR d3
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 809
LOADD d2,8
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
PUSHR d3
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
ADDV d2,7
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 824
LOADD d2,8
PUSHR d2
LOADD d2,5
POP d3
SUBR d3,d2
PUSHR d3
LOADD d2,1
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
SETTRUEI d2
GOTO 905
LOADD d2,2
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
COMPAREV d2,256
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 853
LOADD d2,2
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 873
LOADD d2,2
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,7
PUSHR d2
LOADD d2,3
POP d3
ADDR d2,d3
ADDV d2,7
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0.0001
GOTOMORE 888
LOADD d2,2
PUSHR d2
LOADD d2,5
POP d3
ADDR d2,d3
PUSHR d2
LOADD d2,1
SETR d1,d2
POP d0
ISSOLID d2
CASTBOOLF d2
COMPAREV d2,0.0001
SETMOREI d2
COMPAREV d2,0
SETTRUEI d2
GOTO 905
ALLOCATEMEMV d2,60
STORED d2,0
WRITEPODSTRING d2,"Invalid direction %d passed to CanWalk(x,y,dir,step,bool) \n"
LOADD d2,0
PUSHR d2
LOADD d2,6
PUSHVARGR d2
PRINTFVARG
POP d5
LOADD d3,0
DEALLOCATEMEMR d3
SETV d2,0
GOTO 905
POPARGS d5,0.0009
RETURN
SETR d4,SP
SETR d2,LINKX
ADDV d2,8
GOTO 911
RETURN
SETR d4,SP
SETR d2,LINKY
ADDV d2,8
GOTO 916
RETURN
SETR d4,SP
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,X
PUSHR d2
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,FFTWIDTH
MULTV d2,8
SETV d2,8
PUSHR d2
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,FFTWIDTH
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
GOTO 939
POPARGS d5,0.0001
RETURN
SETR d4,SP
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,Y
PUSHR d2
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,FFTHEIGHT
MULTV d2,8
SETV d2,8
PUSHR d2
LOADD d2,0
PUSHR d2
POP REFFFC
SETR d2,FFTHEIGHT
POP d3
MULTR d2,d3
POP d3
ADDR d2,d3
GOTO 963
POPARGS d5,0.0001
RETURN
PUSHR d2
SETR d4,SP
SETV d2,0
STORED d2,0
LOADD d2,0
COMPAREV d2,14
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 1008
LOADD d2,0
COMPAREV d2,9
SETTRUEI d2
SETV d2,0
COMPAREV d2,0
GOTOTRUE 983
GOTO 1004
LOADD d2,0
COMPAREV d2,6
SETTRUEI d2
SETV d2,0
COMPAREV d2,0
GOTOTRUE 990
GOTO 1004
SETV d2,0
PUSHR d2
LOADD d2,0
PUSHR d2
POP d0
POP d2
SETR BUTTONINPUT,d2
SETV d2,0
PUSHR d2
LOADD d2,0
PUSHR d2
POP d0
POP d2
SETR BUTTONPRESS,d2
LOADD d2,0
ADDV d2,1
STORED d2,0
GOTO 969
POPARGS d5,0.0001
RETURN
SETR d4,SP
LOADD d2,0
SETR d3,d2
COSR d2,d3
PUSHR d2
LOADD d2,1
POP d3
MULTR d2,d3
GOTO 1019
POPARGS d5,0.0002
RETURN
SETR d4,SP
LOADD d2,0
SETR d3,d2
SINR d2,d3
PUSHR d2
LOADD d2,1
POP d3
MULTR d2,d3
GOTO 1030
POPARGS d5,0.0002
RETURN
SETR d4,SP
LOADD d2,0
COMPAREV d2,0
SETLESS d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 1042
SETV d2,1
GOTO 1053
LOADD d2,0
COMPAREV d2,0
SETMORE d2
COMPAREV d2,0
SETTRUEI d2
COMPAREV d2,0
GOTOTRUE 1051
SETV d2,-1
GOTO 1053
SETV d2,0
GOTO 1053
POPARGS d5,0.0001
RETURN


ALLOCATEGMEMV d2,4096
SETR gd1,d2
QUIT